pkg/tfbridge/provider.go

```diff
@@ -309,12 +307,148 @@ func (p *Provider) GetSchema(ctx context.Context,
+	}
+
+	if p.info.PreConfigureCallback != nil {
+		if err = p.info.PreConfigureCallback(news, config); err != nil {
+			return nil, err
+		}
+	}
+
+	// So we can provide better error messages, do a quick scan of required configs for this
+	// schema and report any that haven't been supplied.
+	var missingKeys []*pulumirpc.CheckFailure
+	for key, meta := range p.config {
+		if meta.Required && !config.IsSet(key) {
+			name := TerraformToPulumiName(key, meta, nil, false)
+			fullyQualifiedName := tokens.NewModuleToken(p.pkg(), tokens.ModuleName(name))
+
+			// TF descriptions often have newlines in inopportune positions. This makes them present
+			// a little better in our console output.
+			descriptionWithoutNewlines := strings.Replace(meta.Description, "\n", " ", -1)
+			missingKeys = append(missingKeys, &pulumirpc.CheckFailure{
+				Property: fullyQualifiedName.String(),
+				Reason:   descriptionWithoutNewlines,
+			})
+		}
+	}
+
+	if len(missingKeys) > 0 {
+		return &pulumirpc.CheckResponse{Inputs: req.GetNews(), Failures: missingKeys}, nil
+	}
+
+	// Perform validation of the config state so we can offer nice errors.
+	warns, errs := p.tf.Validate(config)
+	for _, warn := range warns {
+		if err = p.host.Log(ctx, diag.Warning, "", fmt.Sprintf("provider config warning: %v", warn)); err != nil {
+			return nil, err
+		}
+	}
+
+	if len(errs) > 0 {
+		return nil, errors.Wrap(multierror.Append(nil, errs...), "could not validate provider configuration")
+	}
+
+	return &pulumirpc.CheckResponse{Inputs: req.GetNews()}, nil
 }
 
 // DiffConfig diffs the configuration for this Terraform provider.
 func (p *Provider) DiffConfig(ctx context.Context, req *pulumirpc.DiffRequest) (*pulumirpc.DiffResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "DiffConfig is not yet implemented")
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.DiffConfig(%s)", p.label(), urn)
+	glog.V(9).Infof("%s executing", label)
+
+	// There is no logic in the TF provider that suggests that provider level config
+	// should force a new provider. Therefore, we are going to do this based on our
+	// own schema overrides. We should use ForceNew as part of the SchemaInfo to do this
+
+	var olds resource.PropertyMap
+	var err error
+	if req.GetOlds() != nil {
+		olds, err = plugin.UnmarshalProperties(req.GetOlds(), plugin.MarshalOptions{
+			Label: fmt.Sprintf("%s.olds", label), KeepUnknowns: true})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{
+		Label: fmt.Sprintf("%s.news", label), KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	var diffs, replaces []string
+	for _, key := range olds.StableKeys() {
+		// check that the old key exists in the new keys - if not, then report as a diff
+		// check if the new key and the old key are difference
+		if news[key].IsNull() || news[key] != olds[key] {
```

Comment:
Is a reference equality check appropriate for all types of data that might be values of config variables?  For example, for arrays or nested structs? (I believe there are examples of both in AWS for example?)

Reply:
I am more than willing to change this to use reflect.DeepEqual if you think that's the better option here?

Reply:
I’m somewhat nervous that that is not “right” either. Can we ask TF to do this diff for us?  That’s where the actual semantics of diffing these is defined, right?  And aligns with how we do non-config diffs in this provider?  Does the fact that we need to override ForceNew prevent us from doing that?

Reply:
> Can we ask TF to do this diff for us? That’s where the actual semantics of diffing these is defined, right?

If we can do this, it does feel like the right approach. There's a chance that it's not straightforward--we need to build a state and a config to use as inputs to the diff--but I do think it would work. We should be able to override `ForceNew` by post-processing the diff and changing any updates/adds/deletes on force new properties to the corresponding replace. If we can get a TF diff, I _think_ we should also be able to return detailed diffs using the existing code that translates TF diffs to detailed diffs.